### DISCUSSION TITLE
Trigger 404 error manually on page ?

### DISCUSSION THREAD

**Discussion by DevDom05** [nuxt/nuxt#17237] [Questions]
Hi. It's a great version but i have a question. 

I'm test it out and i try to trigger a 404 error 
inside the setup (if a fetch is empty for example or if a fetch don't have spectific data).

How to trigger a 404 error inside the setup() in a page ? :)

**Comment by codeflorist** [comment]
this should work:

```js
const error = new Error(`Page not found: ${useRoute().fullPath}`)
error.statusCode = 404
throw error
```

**Reply by skopz356** [reply]
I tried to use this inside pages component and the error is silenced by Vue.
`
[Vue warn]: Unhandled error during execution of setup function 
  at <[url] onVnodeUnmounted=fn<onVnodeUnmounted> ref=Ref< undefined > key="/kino/testovaci-kinoasd" >
`

**Reply by cawa-93** [reply]
You may use helper:
```ts
throw createError({
  statusCode: 404,
  message: `Page not found: ${useRoute().fullPath}`,
})
```

**Comment by DevDom05** [comment]
Thank you. I will try again ^^

**Comment by Aareksio** [comment]
According to my current knowledge there is [no generic error handler as of right now](https://github.com/nuxt/framework/discussions/559). Neither the custom 404 page works yet. There are 2 different flows for 404 errors:

---

### Server Render
The default Nuxt render function allows us to [throw an error through `ssrContext.error`](https://github.com/nuxt/framework/blob/8056b8e171371795325875facf1735019b9eb475/packages/nitro/src/runtime/app/render.ts#L95-L98). We can use it like so:

```js
const { ssrContext } = useNuxtApp()
ssrContext.error = new Error()
```

This will make our server throw a generic `500 Internal Server Error`. When using nuxt in node.js server mode, the render function is [processed](https://github.com/nuxt/framework/blob/c80870ad8c4a1c1b044d7a5ae27e41e997f8a75f/packages/nitro/src/runtime/server/index.ts) by [h3 web server](https://github.com/unjs/h3). We can adjust our error object to make sure it is [understood by h3](https://github.com/unjs/h3/blob/main/src/error.ts#L14-L18):

```js
const { ssrContext } = useNuxtApp()
ssrContext.error = new Error()
ssrContext.error.statusCode = 404
ssrContext.error.statusMessage= 'Not Found'
```

This is how we currently trigger 404 error with proper status code and status message. Be sure to include the message too, unless you plan to give rest a bad name with `404 Internal Server Error`.

### Client Navigation
This is where things get a bit tricky. Nuxt does not offer proper error handling as of right now, thus your best bet is to handle 404 logic yourself. Create a shared state and render your app conditionally depending on this state (eg. in `app.vue`):

```html
<template>
  <my-error v-if="error"/>
  <nuxt-page v-else/>
</template>

<script setup>
const error = useError()
</script>
```

---

Since you now assign a variable instead of throwing, your components **must** render when fetch fails. Otherwise they may throw (`reading property 'xxx' of undefined`) and your whole app crashes. 

You can (and probably should) create userland abstraction to make dealing with 404s much easier.

---

Happy coding!

**Comment by daniel-fahl** [comment]
For anyone stumbling over this (like myself):
Combining @codeflorist s code and the (now existing) [docs](https://v3.nuxtjs.org/guide/features/error-handling#throwerror) about error handling, I was able to programmatically trigger a 404 error like this:
```js
const error = new Error(`Page not found: ${route.fullPath}`)
error.statusCode = 404
throwError(error);
```

This uses the `throwError()` composable, showing the default 404 error page.
However this does not appear to render a custom `pages/404.vue` file - I might open an issue for that later

**Reply by engenes** [reply]
@danielroe, do you have a solution?

**Reply by danielroe** [reply]
Yes. There is an open PR: https://github.com/nuxt/framework/pull/4539

**Reply by Sitronik** [reply]
> @danielroe, do you have a solution?

I was able to render custom `pages/404` tested only client side without ssr

```js
export default defineNuxtRouteMiddleware(async (to) => {
  try {
    await yourApi.request();
  } catch(e) {
    await router.push('/404');
    window.history.replaceState({}, '', to.path);
  }
});
```
